scipy/stats/_qmc.py

```diff
@@ -481,6 +485,186 @@ def _perturb_discrepancy(sample: np.ndarray, i1: int, i2: int, k: int,
+
+    Parameters
+    ----------
+    sample : array_like (n, d)
+        The sample to iterate on. With ``n`` the number of sample and ``d``
+        the dimension.
+    n_iters : int, optional
+        Number of iterations. Too many iterations tend to cluster the sample
+        as a hypersphere. Default is 10.
+    seed : {None, int, `numpy.random.Generator`}, optional
+        If `seed` is None the `numpy.random.Generator` singleton is used.
+        If `seed` is an int, a new ``Generator`` instance is used,
+        seeded with `seed`.
+        If `seed` is already a ``Generator`` instance then that instance is
+        used.
+
+    Returns
+    -------
+    sample : array_like (n, d)
+        The sample after `n_iters` of Lloyd's algorithm.
+
+    Notes
+    -----
+    Lloyd's algorithm is an iterative process which purpose is to improve
+    the dispersion of a sample. For a given sample: (i) compute a Voronoi
+    Tessellation; (ii) find the centroid of each Voronoi cell; (iii) move the
+    samples toward the centroid of their respective cell.
+
+    The process converges to equally spaced points. It implies that measures
+    like the discrepancy could suffer from too many iterations. On the other
+    hand, L1 and L2 distances should improve. This is especially true with
+    QMC methods which tend to favor the discrepancy over other criteria.
+
+    .. warning::
+
+       The Voronoi Tessellation step is expensive and quickly becomes
+       intractable with dimensions as low as 10 even for a sample of size
+       as low as 1000.
+
+    References
+    ----------
+    .. [1] Lloyd. "Least Squares Quantization in PCM".
+       IEEE Transactions on Information Theory, 1982.
+
+    Examples
+    --------
+    >>> from scipy.spatial.distance import cdist
+    >>> from scipy.stats import qmc
+    >>> rng = np.random.default_rng()
+    >>> sample = rng.random((128, 2))
+
+    Compute the quality of the sample using the L2 criterion.
+
+    >>> def l2_norm(sample):
+    ...    l2 = cdist(sample, sample)
+    ...    return np.min(l2[l2.nonzero()])
+
+    >>> l2_norm(sample)
+    0.00547...  # random
+
+    Now process the sample using Lloyd's algorithm and check the improvement
+    on the L2. The value should increase.
+
+    >>> sample = qmc.lloyd_centroidal_voronoi_tessellation(sample)
+    >>> l2_norm(sample)
+    0.0273...  # random
+
+    """
+    sample = np.asarray(sample)
+    rng = check_random_state(seed)
+
+    if not sample.ndim == 2:
+        raise ValueError('Sample is not a 2D array')
+
+    # Checking that sample is within the bounds
+    if not (np.all(sample >= 0) and np.all(sample <= 1)):
+        raise ValueError('Sample is out of bounds')
+
+    # Fit an exponential to be 2 at 0 and 1 at `n_iters`.
+    # The decay is used for relaxation.
+    # analytical solution for y=exp(-n_iters/x) - 0.1
+    root = -n_iters / np.log(0.1)
+    decay = (np.exp(-x / root) + 0.9 for x in range(n_iters))
```

Comment:
I wonder if we could give users a "switch" to control the relaxation (this may be similar to what CGAL is doing with a convergence criterion?). I don't know if you could have both a sensible default and a way to change it?

I'm just thinking of avoiding user complaints about convergence (we can say "you can tune it yourself").

Reply:
Here I am using a generator, so we  could instead make this an optional parameter. The user would pass a single value or a generator and we would default to the current behaviour.

Reply:
It looks like this has been addressed.